<?xml version="1.0"?>

<launch>


    <arg name="autorun" default="false" />

    <!-- Sequence -->
    <arg name="sequence" default="cloud_avia_mid"/>

    <!-- Bag file -->
    <arg name="lidar_bag_file" value="/home/xy/DATASETS/CTGAUPRO/$(arg sequence)/"/>
                                     
    <!-- Initial start time of each sequence -->
    <arg name="skipped_time" value="4"/>

    <!-- Initial pose in each sequence -->
    <arg name="xyzypr_W_L0" value="[ 0,    0,   0.7,  45,  45, 0,
                                    -0.3, -0.3, 0.5, -135, 0,  0 ]"/>

    <!-- Multi-lidar motion estimation node -->
    <node required="false" pkg="ctgaupro" type="ctgaupro" name="ctgaupro" output="screen"
          launch-prefix="bash -c 'sleep 3; $0 $@' ">
          <!-- launch-prefix="gdb -ex run - -args"> -->

        <!-- Location of the prior map -->
        <param name="priormap_file"   value="/home/xy/DATASETS/CTGAUPRO/priormap.pcd" />

        <!-- Location of bag file -->
        <param name="lidar_bag_file"  value="$(arg lidar_bag_file)" />
        
        <!-- Total number of clouds loaded -->
        <param name="MAX_CLOUDS"      value="-1" />
        <!-- Time since first pointcloud to skip MAP Opt-->
        <param name="SKIPPED_TIME"    value="4.9"/>
        <param name="RECURRENT_SKIP"  value="0" />

        <!-- Set to '1' to skip optimization -->
        <param name="VIZ_ONLY"        value="0"/>
        
        <param name="lidar_topic"     value="['/lidar_0/points', '/lidar_1/points']" type="yaml" />
        <param name="lidar_type"      value="['livox', 'livox']" type="yaml" />
        <param name="stamp_time"      value="['start', 'start']" type="yaml" />

        <!-- Imu topics and their settings -->
        <!-- <param name="imu_topic"       value="['/os_cloud_node/imu']" type="yaml" /> -->

        <!-- Set to 1 to load the search for previous logs and redo -->
        <param name="resume_from_log" value="[0, 0]" type="yaml"/>

        <!-- Initial pose of each lidars-->
        <param name="xyzypr_W_L0"     value="$(arg xyzypr_W_L0)" type="yaml"/>

        <!-- Groundtruth for evaluation-->
        <param name="xtrz_gndtr"      value="[ 0, 0, 0, 0, 0, 0,
                                              -0.1767767, 0, -0.53033009, 180, 45, 0 ]" type="yaml"/>

        <!-- Leaf size to downsample priormap -->
        <param name="pmap_leaf_size"  value="0.15"   />
        <param name="cloud_ds"        value="[0.1]" type="yaml" />
  
        <!-- GN MAP optimization params -->
        <param name="deltaT"          value="0.02204"/>

        <param name="lidar_ds_rate"   value="1"      />
        <param name="lidar_weight"    value="5.0"    />

        <!-- Motion prior factors -->
        <param name="mpSigGa"         value="1.0"    />
        <param name="mpSigNu"         value="1.0"    />

        <!-- Extrinsic factors -->
        <param name="xtSigGa"         value="1.0"    />
        <param name="xtSigNu"         value="1.0"    />

        <!-- Loss function threshold -->
        <param name="ld_loss_thres"   value="-1.0"   />
        <param name="pp_loss_thres"   value="-1.0"   />
        <param name="mp_loss_thres"   value="-10.0"  />
        <param name="xt_loss_thres"   value="-10.0"  />

        <param name="max_omg"         value="-20.0"  />
        <param name="max_alp"         value="-40.0"  />
        <param name="max_vel"         value="-10.0"  />
        <param name="max_acc"         value="-5.0"   />

        <!-- Extrinsic estimation -->
        <param name="SW_CLOUDNUM"     value="60"     />
        <param name="SW_CLOUDSTEP"    value="08"     />
        <param name="max_lidarcoefs"  value="1000"   />
        <param name="XTRZ_DENSITY"    value="1"      />
        <param name="min_planarity"   value="0.5"    />
        <param name="max_plane_dis"   value="0.5"    />
        
        <param name="max_ceres_iter"  value="50"     />
        <param name="max_outer_iter"  value="1"      />
        <param name="min_inner_iter"  value="2"      />
        <param name="max_inner_iter"  value="40"     />
        <param name="conv_thres"      value="2"      />
        <param name="dJ_conv_thres"   value="10"     />
        <param name="conv_dX_thres"   value="[ -0.05, -0.5, -1.0, -0.05, -0.5, -1.0 ]" type="yaml"/>
        <param name="change_thres"    value="[ -15.0, -0.5, -1.0, -15.0, -8.0, -2.0 ]" type="yaml"/>

        <param name="fix_time_begin"  value="0.05"   />
        <param name="fix_time_end"    value="0.0"    />

        <!-- Log dir -->
        <param name="log_period"     value="5.0"     />
        <param name="log_dir" value="/media/tmn/mySataSSD1/Experiments/ctgaupro/logs/lio/sim_$(arg sequence)"/>

    </node>

    <!-- Visualize -->
    <node pkg="rviz" type="rviz" name="rviz_odom" output="log" required="true"
        args="-d $(find ctgaupro)/launch/ctgaupro_lio_sim.rviz" />

</launch>